### PR TITLE
Run trial tests against Python 3.11

### PR DESCRIPTION
**Pull request by DMRobertson**
Part of #13790.

To prove this works, 4f5d492cd6 runs trial against 3.11. **It needs reverting before merging.**

Also, bump dependencies in lockfile to fix a few 3.11 problems:

- [x] Need [wrapt >= 1.14](https://wrapt.readthedocs.io/en/latest/changes.html#version-1-14-0).
- [x] Need Twisted and dependencies release with https://github.com/twisted/twisted/issues/10343
  - currently blocked(?) on https://github.com/glyph/automat/pull/138 being released https://github.com/glyph/automat/issues/139
- [x] Need to work out why [some tests that use PIL fail](https://github.com/matrix-org/synapse/actions/runs/3297558600/jobs/5438633738#step:7:3648)
  - There was no 3.11 wheel for the pinned pillow version. Seems to work after an update.
- [x] Need to investigate unexplained [failure in ijson closing coroutines test](https://github.com/matrix-org/synapse/actions/runs/3297558600/jobs/5438633738#step:7:3744)
  - Error in test. Dunno why it shows up now, but easily fixed.

### CODE REVIEW

**Comment by DMRobertson** [comment]
I can't reproduce the failures in `tests.rest.media.v1.test_media_storage`.

There are failures in some kind of logging test which reaches into twisted and automat internals?!

```
Traceback (most recent call last):
  File "/home/runner/work/synapse/synapse/tests/logging/test_remote_handler.py", line 158, in test_cancel_connection
    handler = RemoteHandler(
  File "/home/runner/work/synapse/synapse/synapse/logging/_remote.py", line 142, in __init__
    self._service.startService()
  File "/home/runner/.cache/pypoetry/virtualenvs/matrix-synapse-pswDeSvb-py3.11/lib/python3.11/site-packages/twisted/application/internet.py", line 1178, in startService
    self._machine.start()
  File "/home/runner/.cache/pypoetry/virtualenvs/matrix-synapse-pswDeSvb-py3.11/lib/python3.11/site-packages/automat/_methodical.py", line 232, in __get__
    @preserveName(self.method)
  File "/home/runner/.cache/pypoetry/virtualenvs/matrix-synapse-pswDeSvb-py3.11/lib/python3.11/site-packages/automat/_introspection.py", line 43, in decorator
    return copyfunction(decorated,
  File "/home/runner/.cache/pypoetry/virtualenvs/matrix-synapse-pswDeSvb-py3.11/lib/python3.11/site-packages/automat/_introspection.py", line 35, in copyfunction
    return function(copycode(template.__code__, codechanges), *values)
  File "/home/runner/.cache/pypoetry/virtualenvs/matrix-synapse-pswDeSvb-py3.11/lib/python3.11/site-packages/automat/_introspection.py", line 23, in copycode
    return code(*values)
builtins.TypeError: code() argument 13 must be str, not int

tests.logging.test_remote_handler.RemoteHandlerTestCase.test_cancel_connection
```

Also an unexplained timeout:

```
raceback (most recent call last):
  File "/home/runner/work/synapse/synapse/tests/rest/admin/test_server_notice.py", line 175, in test_invalid_avatar_url
    channel = self.make_request(
  File "/home/runner/work/synapse/synapse/tests/unittest.py", line 495, in make_request
    return make_request(
  File "/home/runner/work/synapse/synapse/tests/server.py", line 387, in make_request
    channel.await_result()
  File "/home/runner/work/synapse/synapse/tests/server.py", line 240, in await_result
    raise TimedOutException("Timed out waiting for request to finish.")
tests.server.TimedOutException: Timed out waiting for request to finish.

tests.rest.admin.test_server_notice.ServerNoticeTestCase.test_invalid_avatar_url
-------------------------------------------------------------------------------
Ran 2473 tests in 1037.253s
```

Relevant test output:

<details>

```
--> tests.rest.admin.test_server_notice.ServerNoticeTestCase.test_invalid_avatar_url <--
  2022-09-14 16:41:23,238 - synapse.http.site - 302 - ERROR - POST-2212 - Asynchronous message handler raised an uncaught exception
  Traceback (most recent call last):
    File "/home/runner/.cache/pypoetry/virtualenvs/matrix-synapse-pswDeSvb-py3.11/lib/python3.11/site-packages/twisted/internet/defer.py", line 1660, in _inlineCallbacks
      result = current_context.run(gen.send, result)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  StopIteration
  
  During handling of the above exception, another exception occurred:
  
  Traceback (most recent call last):
    File "/home/runner/.cache/pypoetry/virtualenvs/matrix-synapse-pswDeSvb-py3.11/lib/python3.11/site-packages/twisted/internet/defer.py", line 1660, in _inlineCallbacks
      result = current_context.run(gen.send, result)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  StopIteration: <FrozenEventV3 event_id=$HeN5M_P9r_kGWqdqfCZ4YWcpE5SYxjSo2eeGWOC--iA, type=m.room.create, state_key=, outlier=False>
  
  During handling of the above exception, another exception occurred:
  
  Traceback (most recent call last):
    File "/home/runner/work/synapse/synapse/synapse/http/server.py", line 306, in _async_render_wrapper
      callback_return = await self._async_render(request)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/home/runner/work/synapse/synapse/synapse/http/server.py", line 512, in _async_render
      callback_return = await raw_callback_return
                        ^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/home/runner/work/synapse/synapse/synapse/rest/admin/server_notice_servlet.py", line 99, in on_POST
      event = await self.server_notices_manager.send_notice(
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/home/runner/work/synapse/synapse/synapse/server_notices/server_notices_manager.py", line 68, in send_notice
      room_id = await self.get_or_create_notice_room_for_user(user_id)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  ValueError: mxc URI 'somthingwrong' did not match expected format
  
  During handling of the above exception, another exception occurred:
  
  Traceback (most recent call last):
    File "/home/runner/work/synapse/synapse/synapse/http/site.py", line 299, in processing
      yield
    File "/home/runner/work/synapse/synapse/synapse/http/server.py", line 228, in wrapped_async_request_handler
      await h(self, request)
    File "/home/runner/work/synapse/synapse/synapse/http/server.py", line 316, in _async_render_wrapper
      self._send_error_response(f, request)
    File "/home/runner/work/synapse/synapse/synapse/http/server.py", line 524, in _send_error_response
      return_json_error(f, request, self.hs.config)
    File "/home/runner/work/synapse/synapse/synapse/http/server.py", line 123, in return_json_error
      logger.error(
    File "/opt/hostedtoolcache/Python/3.11.0-rc.2/x64/lib/python3.11/logging/__init__.py", line [1518](https://github.com/matrix-org/synapse/actions/runs/3054423174/jobs/4926379440#step:8:1519), in error
      self._log(ERROR, msg, args, **kwargs)
    File "/opt/hostedtoolcache/Python/3.11.0-rc.2/x64/lib/python3.11/logging/__init__.py", line 1634, in _log
      self.handle(record)
    File "/opt/hostedtoolcache/Python/3.11.0-rc.2/x64/lib/python3.11/logging/__init__.py", line 1644, in handle
      self.callHandlers(record)
    File "/opt/hostedtoolcache/Python/3.11.0-rc.2/x64/lib/python3.11/logging/__init__.py", line 1706, in callHandlers
      hdlr.handle(record)
    File "/opt/hostedtoolcache/Python/3.11.0-rc.2/x64/lib/python3.11/logging/__init__.py", line 978, in handle
      self.emit(record)
    File "/home/runner/work/synapse/synapse/tests/test_utils/logging_setup.py", line 28, in emit
      log_entry = self.format(record)
                  ^^^^^^^^^^^^^^^^^^^
    File "/opt/hostedtoolcache/Python/3.11.0-rc.2/x64/lib/python3.11/logging/__init__.py", line 953, in format
      return fmt.format(record)
             ^^^^^^^^^^^^^^^^^^
    File "/opt/hostedtoolcache/Python/3.11.0-rc.2/x64/lib/python3.11/logging/__init__.py", line 695, in format
      record.exc_text = self.formatException(record.exc_info)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/opt/hostedtoolcache/Python/3.11.0-rc.2/x64/lib/python3.11/logging/__init__.py", line 645, in formatException
      traceback.print_exception(ei[0], ei[1], tb, None, sio)
    File "/opt/hostedtoolcache/Python/3.11.0-rc.2/x64/lib/python3.11/traceback.py", line 124, in print_exception
      te = TracebackException(type(value), value, tb, limit=limit, compact=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/opt/hostedtoolcache/Python/3.11.0-rc.2/x64/lib/python3.11/traceback.py", line 688, in __init__
      self.stack = StackSummary._extract_from_extended_frame_gen(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/opt/hostedtoolcache/Python/3.11.0-rc.2/x64/lib/python3.11/traceback.py", line 416, in _extract_from_extended_frame_gen
      for f, (lineno, end_lineno, colno, end_colno) in frame_gen:
    File "/opt/hostedtoolcache/Python/3.11.0-rc.2/x64/lib/python3.11/traceback.py", line 353, in _walk_tb_with_full_positions
      positions = _get_code_position(tb.tb_frame.f_code, tb.tb_lasti)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/opt/hostedtoolcache/Python/3.11.0-rc.2/x64/lib/python3.11/traceback.py", line 366, in _get_code_position
      positions_gen = code.co_positions()
                      ^^^^^^^^^^^^^^^^^
  AttributeError: '_Code' object has no attribute 'co_positions'
```
</details>

**Comment by DMRobertson** [comment]
>   File "/home/runner/.cache/pypoetry/virtualenvs/matrix-synapse-pswDeSvb-py3.11/lib/python3.11/site-packages/automat/_introspection.py", line 23, in copycode
    return code(*values)
> builtins.TypeError: code() argument 13 must be str, not int

Would guess that https://github.com/glyph/automat/commit/d0c2845ead9b8f576042d42134bfb5c63086be39 or https://github.com/twisted/twisted/pull/1723 will help in this area.

**Comment by DMRobertson** [comment]
> I can't reproduce the failures in `tests.rest.media.v1.test_media_storage`.

Only MediaRepoTests_2 fails, which I think corresponds to a webp example. It might be that the version of Pillow installed on the runner lacks webp support?

**Comment by DMRobertson** [comment]
Pip doesn't know how to verify hashes for dependencies defined in git:

```
   #22 2.582 ERROR: Can't verify hashes for these requirements because we don't have a way to hash version control repositories:
  #22 2.582     automat@ git+https://github.com/glyph/automat.git@4260ebbc3347b12566f219fe0e887c18bf972cb9 from git+https://github.com/glyph/automat.git@4260ebbc3347b12566f219fe0e887c18bf972cb9 (from -r /synapse/requirements.txt (line 7))
  #22 2.657 WARNING: You are using pip version 22.0.4; however, version 22.3 is available.
  #22 2.657 You should consider upgrading via the '/usr/local/bin/python -m pip install --upgrade pip' command.
  #22 ERROR: executor failed running [/bin/sh -c pip install --prefix="/install" --no-deps --no-warn-script-location -r /synapse/requirements.txt]: exit code: 1
  ------
   > [builder  6/10] RUN --mount=type=cache,target=/root/.cache/pip   pip install --prefix="/install" --no-deps --no-warn-script-location -r /synapse/requirements.txt:
```

which is a PITA

File: poetry.lock [hunk 1]
```diff
@@ -671,12 +671,16 @@ python-versions = "*"
 
 [[package]]
 name = "pillow"
-version = "9.0.1"
+version = "9.2.0"
```
Comment (DMRobertson) [inline poetry.lock:674]: This seemed to be needed to get 3.11 wheels.

File: poetry.lock [hunk 2]
```diff
@@ -1533,7 +1537,7 @@ python-versions = "*"
 
 [[package]]
 name = "wrapt"
-version = "1.13.3"
+version = "1.14.1"
```
Comment (DMRobertson) [inline poetry.lock:1540]: Old version was using a deprecated function which was removed in Python 3.11, see https://github.com/GrahamDumpleton/wrapt/pull/205

File: poetry.lock [hunk 3]
```diff
@@ -25,7 +25,7 @@ cryptography = ">=3.2"
 
 [[package]]
 name = "automat"
-version = "20.2.0"
+version = "22.10.0"
```
Comment (DMRobertson) [inline poetry.lock:28]: Needed for https://github.com/glyph/automat/issues/135

**Review by squahtx — APPROVED** [review]
LGTM if CI is happy!

**Comment by DMRobertson** [comment]
> To prove this works, [4f5d492](https://github.com/matrix-org/synapse/commit/4f5d492cd6a9438de03d1b768f4c220cb662ac06) runs trial against 3.11. **It needs reverting before merging.**

OOPS

**Comment by DMRobertson** [comment]
The release branch is failing to `setup-python-poetry`, see e.g. https://github.com/matrix-org/synapse/actions/runs/3389272009/jobs/5632193743. It seems that poetry running on CPython 3.11 is unable to install wrapt 1.13.3. I'm going to try cherry picking this (plus the fixup commit) to the release branch.

**Pull request merged**